### ISSUE TITLE
carbon-cache update, now it is blocking when Tagging !!!

### ISSUE THREAD

**Issue by xmdevops** [graphite-project/carbon#737]
[root@23-91-102-27 bin]# python carbon-cache.py --nodaemon --debug start
Removing stale pidfile /opt/graphite/storage/carbon-cache-a.pid
Starting carbon-cache (instance a)
17/01/2018 14:06:18 :: [console] Using sorted write strategy for cache
17/01/2018 14:06:18 :: [console] Enabling Whisper fallocate support
17/01/2018 14:06:18 :: [console] twistd 17.9.0 (/usr/bin/python 2.7.5) starting up.
17/01/2018 14:06:18 :: [console] reactor class: twisted.internet.epollreactor.EPollReactor.
17/01/2018 14:06:18 :: [console] ServerFactory starting on 7002
17/01/2018 14:06:18 :: [console] Starting factory <twisted.internet.protocol.ServerFactory instance at 0x161bcf8>
17/01/2018 14:06:18 :: [console] Installing SIG_IGN for SIGHUP
17/01/2018 14:06:18 :: [console] CarbonReceiverFactory starting on 2003
17/01/2018 14:06:18 :: [console] Starting factory <carbon.protocols.CarbonReceiverFactory instance at 0x15304d0>
17/01/2018 14:06:18 :: [console] CarbonReceiverFactory starting on 2004
17/01/2018 14:06:18 :: [console] Starting factory <carbon.protocols.CarbonReceiverFactory instance at 0x1595ef0>
17/01/2018 14:06:30 :: [listener] MetricLineReceiver connection with 183.129.198.242:4234 established
17/01/2018 14:06:30 :: [listener] MetricLineReceiver connection with 183.129.198.242:4234 closed cleanly
17/01/2018 14:06:31 :: [tagdb] Tagging local.random.diceroll
17/01/2018 14:06:31 :: [tagdb] Tagged local.random.diceroll in 0.0339570045471

It's been blocked all the time~

**Comment by deniszh** [comment]
Could you please elaborate your issue? What you mean saying "blocked"?
Logs above looks perfectly fine.

**Comment by xmdevops** [comment]
I'm glad you responded so quickly， i use the master branch, and it blocked and nothing printed, i use the nc test like this 
`echo "local.random.diceroll 4 `date +%s`"|nc -c 10.2.5.51 2003`

now i test the branch 1.0.x, everything is ok now, and i will test 1.1.x later

i  like your project very much, and I'm developing two times based on your project now

**Comment by xmdevops** [comment]
when test with nc ok, carbon-cache should print like this
`17/01/2018 17:21:04 :: [listener] MetricLineReceiver connection with 183.129.198.242:5460 established
17/01/2018 17:21:04 :: [listener] MetricLineReceiver connection with 183.129.198.242:5460 closed cleanly
17/01/2018 17:21:05 :: [listener] MetricLineReceiver connection with 183.129.198.242:5461 established
17/01/2018 17:21:05 :: [listener] MetricLineReceiver connection with 183.129.198.242:5461 closed cleanly
`
master branch blocked and nothing printed, is the twisted version issue?

**Comment by deniszh** [comment]
Sorry, @xmdevops, I'm still not understanding what's your issue is, what's exactly blocked and what's not working.
nc stucks when submitting metrics? Then it's something really wrong with Twisted, indeed, but I never saw anything similar.

**Comment by xmdevops** [comment]
i test the branch 1.1.x, found the same issue, may be this days you merge the new code, issue is that:
1. run python carbon-cache.py --nodaemon --debug start
2. echo "local.random.diceroll 4date +%s"|nc -c 10.2.5.51 2003
3. and then the carbon-cache is blocked, can not handler any request, the nc wait timeout
4. because the carbon-cache is blocked（i dont know why?），the series data can not be stored~ 

my env:
twistd 16.4.1 
python 2.7.5

and the previous 1.0.x not found GRAPHITE_URL in carbon.conf, not used for tagging? so have not the issue? 
 
notes: three weeks ago i use the master branch, everything is working ok, today i redeploy all compant, and the issue i meet~ 

looking forward to helping to solve this problem

**Comment by deniszh** [comment]
Tried to reproduce - have no issue on Twisted 17.9.0
```
17/01/2018 14:23:03 :: [console] Using sorted write strategy for cache
17/01/2018 14:23:03 :: [console] 'WHISPER_FALLOCATE_CREATE is enabled but linking failed.'
17/01/2018 14:23:03 :: [console] /Users/dzhdanov/.graphite/conf/storage-aggregation.conf not found or wrong perms, ignoring.
17/01/2018 14:23:03 :: [console] twistd 17.9.0 (/Users/dzhdanov/.graphite/bin/python2 2.7.14) starting up.
17/01/2018 14:23:03 :: [console] reactor class: twisted.internet.selectreactor.SelectReactor.
17/01/2018 14:23:03 :: [console] ServerFactory starting on 7002
17/01/2018 14:23:03 :: [console] Starting factory <twisted.internet.protocol.ServerFactory instance at 0x10dec9758>
17/01/2018 14:23:03 :: [console] Installing SIG_IGN for SIGHUP
17/01/2018 14:23:03 :: [console] CarbonReceiverFactory starting on 2003
17/01/2018 14:23:03 :: [console] Starting factory <carbon.protocols.CarbonReceiverFactory instance at 0x10ddccd88>
17/01/2018 14:23:03 :: [console] CarbonReceiverFactory starting on 2004
17/01/2018 14:23:03 :: [console] Starting factory <carbon.protocols.CarbonReceiverFactory instance at 0x10dedc7e8>
17/01/2018 14:23:07 :: [listener] MetricLineReceiver connection with 127.0.0.1:65483 established
17/01/2018 14:23:07 :: [listener] invalid line received from client 127.0.0.1:65483, ignoring [local.random.diceroll 4 ( +%s)]
17/01/2018 14:23:07 :: [listener] MetricLineReceiver connection with 127.0.0.1:65483 closed cleanly
17/01/2018 14:23:40 :: [listener] MetricLineReceiver connection with 127.0.0.1:65501 established
17/01/2018 14:23:40 :: [listener] MetricLineReceiver connection with 127.0.0.1:65501 closed cleanly
17/01/2018 14:23:40 :: [tagdb] Tagging foo.bar, foo.bar
17/01/2018 14:23:41 :: [tagdb] Error tagging foo.bar, foo.bar: Connection was refused by other side: 61: Connection refused.
17/01/2018 14:23:41 :: [console] Unhandled Error
Traceback (most recent call last):
Failure: twisted.internet.error.ConnectionRefusedError: Connection was refused by other side: 61: Connection refused.
```
Probably relatest with latest commits. Soon we'll try to make some refactoring there.
Please also note that master is not intended to be used in production, please use stable branch (1.1.x now).

**Comment by deniszh** [comment]
Ah, probably it's because some commits which are backported to 1.1.x
Like https://github.com/graphite-project/carbon/commit/d17f87d3617862b28257a104398172044065e07b or https://github.com/graphite-project/carbon/commit/cca9aa37bf96a6e8d039e16339fb4f7627b79e98 or both.

Did you try 1.1.1 release? or only branch?

**Comment by DanCech** [comment]
@xmdevops can you try https://github.com/graphite-project/carbon/pull/738 ?

**Comment by xmdevops** [comment]
@deniszh i try only git clone -b  1.1.x https://github.com/graphite-project/carbon.git, i found the issue with this compant

**Comment by xmdevops** [comment]
@DanCech #738  is ok, everything is work well ~ thanks very much，but i have some other question
1. i want use ceres replace whisper, but i found your some compant like carbon graphite-web band the whisper, i dont know how to replace~

**Comment by DanCech** [comment]
graphite-web supports ceres, you just need to configure STORAGE_FINDERS http://graphite.readthedocs.io/en/latest/storage-backends.html and possibly CERES_DIR http://graphite.readthedocs.io/en/latest/config-local-settings.html if you aren't using the default directory.

**Comment by deniszh** [comment]
Issue fixed in master and 1.1.x, going to make a release real soon.
